modules/generator/processor/localblocks/processor.go

```diff
@@ -276,16 +279,36 @@ func (p *Processor) flushLoop() {
 
 		op := o.(*flushOp)
 		op.attempts++
+
+		if op.attempts > maxFlushAttempts {
+			_ = level.Error(p.logger).Log("msg", "failed to flush block after max attempts", "tenant", p.tenant, "block", op.blockID, "attempts", op.attempts)
+
+			// attempt to delete the block
+			p.blocksMtx.Lock()
+			err := p.wal.LocalBackend().ClearBlock(op.blockID, p.tenant)
+			if err != nil {
+				_ = level.Error(p.logger).Log("msg", "failed to clear corrupt block", "tenant", p.tenant, "block", op.blockID, "err", err)
+			}
+			delete(p.completeBlocks, op.blockID)
+			p.blocksMtx.Unlock()
+
+			continue
+		}
+
 		err := p.flushBlock(op.blockID)
 		if err != nil {
-			_ = level.Error(p.logger).Log("msg", "failed to flush a block", "err", err)
-
 			_ = level.Info(p.logger).Log("msg", "re-queueing block for flushing", "block", op.blockID, "attempts", op.attempts)
-			op.at = time.Now().Add(op.backoff())
+			metricFailedFlushes.Inc()
 
-			if _, err := p.flushqueue.Enqueue(op); err != nil {
-				_ = level.Error(p.logger).Log("msg", "failed to requeue block for flushing", "err", err)
-			}
+			delay := op.backoff()
+			op.at = time.Now().Add(delay)
+
+			go func() {
+				time.Sleep(delay)
+				if _, err := p.flushqueue.Enqueue(op); err != nil {
+					_ = level.Error(p.logger).Log("msg", "failed to requeue block for flushing", "err", err)
```

Comment:
What happens if we fail to requeue? Should we delete the block as well?

Reply:
i think no. the only time enqueue can fail is if the queue is stopped:

https://github.com/grafana/tempo/blob/main/pkg/flushqueues/priority_queue.go#L91-L93

in this case the process is shutting down and so we want to keep the block for when it starts back up. we could definitely be more explicit about this tho.